### ISSUE TITLE
Does quagga JS work with GS1-128 (UCC/EAN-128) format?

### ISSUE THREAD

**Issue by mlkae** [serratus/quaggaJS#357]
I tried to set readers as ean_reader but it doesn't work. GS1-128 (UCC/EAN-128)  format is supported? In this case, how is the correct readers setting?

Thank you.

**Comment by getsalty** [comment]
After a little research, I found this article:
https://www.keyence.com/ss/products/auto_id/barcode_lecture/basic/code128/
.

It states "Since GS1-128 uses the barcode composition of CODE 128, any barcode reader that can read CODE 128 can be used to read the GS1-128 data."

Quagga can read code 128 barcodes (in fact I think its the 'default' barcode) so it should work fine for GS1-128 barcodes.  You will probably need to apply your own logic/parsing once the barcode is read to determine AI-Value relationships, but luckily regex makes that easy.

To get started, use examples in the readme and the examples folder in the source code.  Make sure to use code 128.

**Comment by mlkae** [comment]
I tried in this way:
      Quagga.decodeSingle({
                    decoder: {                       
                        readers: ["code_128_reader"]                      
                    },
                    locate: true, // try to locate the barcode in the image
                    multiple: true,                                  
                      src: '../Content/img/code_128.png'                 
                 
                }, function (result) {
                    if (result.codeResult) {
                        alert(result.codeResult.code);
                    } else {
                        alert("barcode not detected");
                    }
                });

but the "result" variable is undefined. 
Just to test, I tried to set  readers: ["ean_reader"] and "results"contains something, but only "boxes" data.

I tried with a test imaget (in the attachment)  that I got here https://barcode.tec-it.com/it/Code128?data=12345678 
setting Code-128 and  '12345678' as string.
I thought to generate also a GS1 code (that is what I need to decode in my application) but I have problem with Code 128 too.... I'm doing something wrong?

Thank you.



![code_128](https://user-images.githubusercontent.com/48021937/53506967-591cda00-3ab7-11e9-941c-0d9579998ffb.png)

**Comment by getsalty** [comment]
I made this to help you out:
https://jsfiddle.net/xelfarcherdv/8cjgqo4h/

Not my prettiest work, but will do.
Using the website you linked to, I created the default GS1-128 barcode: https://imgur.com/9RQM28z

Notice that I am running configuration for code 128.

**Comment by mlkae** [comment]
I think that you used the GS1 Composite Simbology instead of GS1-128 (UCC/EAN-128)  but I tried with the GS1-128 (UCC/EAN-128 default one and your example works, thank you!
I'll try to adapt my code according to your example even if in our application we dont' use  React  and we cannot use ES6 ( customers ask  for IE compatibility and using Babel now would require too much effort).   Maybe my configuration of quagga was not correct .... :(
By the way, I have to try diffrerent settings everytime before getting the right decoding configuration?

Thank you for your help.

**Comment by mlkae** [comment]
I think there is a problem. I tried to run your example many times now with 2 GS1-128 (UCC/EAN-128) barcode but it gives me  inconsistent results : most of the time FAILED, sometimes green, but it seems that the data are stil linked to the first image. If I try with my image first, it gives always FAILED. Could you try with one of the images in attachment? Maybe it's a cache problem.

![gs1_default](https://user-images.githubusercontent.com/48021937/53567109-81abdf00-3b5e-11e9-8ebe-d2c361be4c9c.png)

![gs1_def](https://user-images.githubusercontent.com/48021937/53567117-8a041a00-3b5e-11e9-9206-70f60da72732.png)

I tried with my updated code, i set one of the "right" settings  and it works only with your image (GS1 Composite Simbology) but doesn't work with GS1-128 (UCC/EAN-128).

Quagga.decodeSingle({
                    locator: {
                        patchSize: "medium",
                        halfSample:  0
                    },
                    decoder: {
                        readers: [
                          {
                              format: "code_128_reader",
                              config: {}
                          }
                        ],
                        debug: {
                            drawBoundingBox: false,
                            showFrequency: false,
                            drawScanline: true,
                            showPattern: false
                        }
                    },
                    inputStream: { size: 1280 },
                    locate: true, // try to locate the barcode in the image
                    multiple: true,            
                    src: '../Content/img/gs1.png'
                
                }, function (result) {
                    if (result.codeResult) {
                        alert(result.codeResult.code);
                    } else {
                        alert("barcode not detected");
                    }
                });




![screen](https://user-images.githubusercontent.com/48021937/53572937-b1fa7a00-3b6c-11e9-91c8-3c179d62f738.png)

**Comment by getsalty** [comment]
Don't have too much time to respond, but let me try to explain a few things.

My example runs through a bunch of different quagga configurations and tells you if your device can read the barcode using those configurations.  So if a configuration succeeds, then you can use those values when you are initializing Quagga and it should work for you.

Also, the issue with your images is probably because there is no white-space on the sides of the barcodes.  I have found that having some sort of space around the barcode helps a ton.  There is space around the barcode in real life, so make sure the pictures also have space around it (my linked image has space around it).  When I copied your barcodes into Paint and made sure there was space around it, the scanner was able to detect each barcode easily.

**Comment by ericblade** [comment]
ah, yes, you absolutely have to have enough space around the barcode to be able to handle it -- a scanner reads the white space not the black space, is my understanding.

**Comment by mlkae** [comment]
Yes, I'm trying.. but the image has to be "perfect" to be decoded, I mean white spaces... opacque paper.. alone barcode... it's not so easy :) I'm wondering if zxing library in .NET could help instead of working on client side
